extensions/wikia/UserRenameTool/SpecialRenameuser_body.php

```diff
@@ -32,99 +31,183 @@ public function execute( $par ) {
+
+		if ( $requestorUser->isAnon() || !$requestorUser->isAllowed( 'renameuser' ) ) {
+			throw new PermissionsError( 'renameuser' );
+		}
+
+		if ( $username && !$requestorUser->isAllowed( 'renameanotheruser' ) ) {
+			throw new PermissionsError( 'renameanotheruser' );
+		}
 
 		if ( wfReadOnly() || !$wgStatsDBEnabled ) {
-			$out->readOnlyPage();
+			$this->getOutput()->readOnlyPage();
+
 			return;
 		}
 
-		// Get the request data
-		$request = $this->getRequest();
-		$oldUsername = $request->getText( 'oldusername', $par );
-		$newUsername = $request->getText( 'newusername' );
-		$reason = $request->getText( 'reason' );
-		$token = $this->getUser()->getEditToken();
-		$notifyRenamed = $request->getBool( 'notify_renamed', false );
-		$confirmAction = false;
+		// TODO: remove after QA tests
+		if ( self::shouldUnlock() ) {
+			$requestorUser->setGlobalFlag( 'wasRenamed', false );
+			$requestorUser->saveSettings();
+		}
+		// TODO: end
 
-		if ( $request->wasPosted() && $request->getInt( 'confirmaction' ) ) {
-			$confirmAction = true;
+		if ( \RenameUserProcess::canUserChangeUsername( $requestorUser ) ) {
+			$this->renderForm( $requestorUser, $username );
+		} else {
+			$this->renderDisallow();
 		}
 
-		$warnings = [];
-		$errors = [];
-		$info = [];
-
-		if ( $this->validRenameRequest() ) {
-			$process = new RenameUserProcess( $oldUsername, $newUsername, $confirmAction, $reason, $notifyRenamed );
-			$status = $process->run();
-			$warnings = $process->getWarnings();
-			$errors = $process->getErrors();
-			if ( $status ) {
-				$info[] = $this->msg( 'userrenametool-info-in-progress' )->inContentLanguage()->escaped();
-			}
+		return;
+	}
+
+	public static function validateData( array $data, User $user, bool $selfRename = true ) {
+		$errorList = [];
+		$newUsername = $data['newUsername'];
+
+		if ( $data['token'] === '' ) {
+			$errorList[] = 'userrenametool-error-token-not-exists';
 		}
 
-		$showConfirm = ( empty( $errors ) && empty( $info ) );
+		if ( $newUsername === '' ) {
+			$errorList[] = 'userrenametool-error-no-username';
+		}
 
-		// note: errors and info beyond this point are non-blocking
+		if ( $newUsername !== $data['newUsernameRepeat' ] ) {
+			$errorList[] = 'userrenametool-error-not-repeated-correctly';
+		}
 
-		if ( !empty( $oldUsername ) ) {
-			$oldUser = User::newFromName( $oldUsername );
-			if ( $oldUser->getGlobalFlag( 'requested-rename', 0 ) ) {
-				$info[] = $this->msg( 'userrenametool-requested-rename', $oldUsername )->escaped();
-			} else {
-				$errors[] = $this->msg( 'userrenametool-did-not-request-rename', $oldUsername )->escaped();
-			}
-			if ( $oldUser->getGlobalFlag( 'wasRenamed', 0 ) ) {
-				$errors[] = $this->msg( 'userrenametool-previously-renamed', $oldUsername )->escaped();
+		if ( !\User::isValidUserName( $newUsername ) ) {
+			$errorList[] = 'userrenametool-error-non-alphanumeric';
+		}
+
+		if ( strlen( $newUsername ) > \RenameUserProcess::MAX_USERNAME_LENGTH ) {
```

Comment:
The name can have UTF characters. Should we use `mb_strlen` instead?

Reply:
Ideally I think such validation should be centralized.